File: src/test/java/lan/dk/podcastserver/business/CoverBusinessTest.java
package lan.dk.podcastserver.business;

import com.github.tomakehurst.wiremock.junit.WireMockRule;
import lan.dk.podcastserver.entity.Cover;
import lan.dk.podcastserver.entity.CoverAssert;
import lan.dk.podcastserver.entity.Podcast;
import lan.dk.podcastserver.repository.CoverRepository;
import lan.dk.podcastserver.service.PodcastServerParameters;
import lan.dk.podcastserver.service.UrlService;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Spy;
import org.mockito.runners.MockitoJUnitRunner;
import org.springframework.util.FileSystemUtils;

import java.net.URI;
import java.net.URISyntaxException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.*;

/**
 * Created by kevin on 24/07/15 for Podcast Server
 */
@RunWith(MockitoJUnitRunner.class)
public class CoverBusinessTest {

    String ROOT_FOLDER = "/tmp/podcast";

    @Rule
    public WireMockRule wireMockRule = new WireMockRule(8089); // No-args constructor defaults to port 8080

    @Before
    public void beforeEach() {
        FileSystemUtils.deleteRecursively(Paths.get(ROOT_FOLDER).toFile());
    }

    @Mock CoverRepository coverRepository;
    @Mock PodcastServerParameters podcastServerParameters;
    @Spy  UrlService urlService;
    @InjectMocks CoverBusiness coverBusiness;

    @Test
    public void should_find_one() {
        /* Given */
        when(coverRepository.findOne(any(UUID.class))).then(invocation -> new Cover().setId((UUID) invocation.getArguments()[0]));
        UUID id = UUID.randomUUID();
        /* When */
        Cover cover = coverBusiness.findOne(id);
        /* Then */  CoverAssert.assertThat(cover).hasId(id);
        verify(coverRepository, times(1)).findOne(eq(id));
    }

    @Test
    public void should_say_it_same_cover() {
        /* Given */ Podcast p1 = new Podcast(), p2 = new Podcast();
        p1.setCover(new Cover("http://fakeUrl.com"));
        p2.setCover(new Cover("http://fakeurl.com"));

        /* When */
        Boolean isSame = coverBusiness.hasSameCoverURL(p1, p2);
        /* Then */ assertThat(isSame).isTrue();
    }

    @Test
    public void should_say_its_not_same_cover() {
        /* Given */ Podcast p1 = new Podcast(), p2 = new Podcast();
        p1.setCover(new Cover("http://fakeUrl.com"));
        p2.setCover(new Cover("https://fakeurl.com"));

        /* When */
        Boolean isSame = coverBusiness.hasSameCoverURL(p1, p2);
        /* Then */ assertThat(isSame).isFalse();
    }

    @Test
    public void should_not_treat_this_podcast() {
        /* Given */ Podcast podcast = new Podcast();
        /* When */  String url = coverBusiness.download(podcast);
        /* Then */  assertThat(url).isEmpty();
    }

    @Test
    public void should_download_the_cover() throws URISyntaxException {
        /* Given */
        String podcastTitle = "Foo";
        String imageExtension = "png";
        String defaultCoverValue = "cover";
        String serverUrl = "http://localhost:8080/";
        Podcast podcast = new Podcast().setTitle(podcastTitle).setId(UUID.randomUUID());
        podcast.setCover(new Cover("http://localhost:8089/img/image." + imageExtension));
        when(podcastServerParameters.coverDefaultName()).thenReturn(defaultCoverValue);
        when(podcastServerParameters.rootFolder()).thenReturn(Paths.get(ROOT_FOLDER));
        when(podcastServerParameters.serverUrl()).thenReturn(new URI(serverUrl));
        /* When */
        String url = coverBusiness.download(podcast);

        /* Then */
        assertThat(url).isEqualTo(serverUrl + "api/podcast/" + podcast.getId() + "/" + defaultCoverValue + "." + imageExtension);
    }

    @Test
    public void should_reject_by_network_exception() throws URISyntaxException {
        /* Given */
        Podcast podcast = new Podcast().setTitle("Foo");
        podcast.setCover(new Cover("http://localhost:8089/img/image.jpg"));
        when(podcastServerParameters.coverDefaultName()).thenReturn("cover");
        when(podcastServerParameters.rootFolder()).thenReturn(Paths.get("/tmp"));
        when(podcastServerParameters.fileContainer()).thenReturn(new URI("http://localhost:8080/podcast"));
        /* When */
        String url = coverBusiness.download(podcast);

        /* Then */
        assertThat(url).isEmpty();
    }

    @Test
    public void should_generate_path_cover_of_podcast() {
        /* Given */
        Podcast podcast = Podcast.builder()
                .id(UUID.randomUUID())
                .title("A Podcast")
                .cover(new Cover("http://podcast.dk.lan/podcast/Google/aCover.jpg"))
                .build();

        when(podcastServerParameters.coverDefaultName()).thenReturn("cover");
        when(podcastServerParameters.rootFolder()).thenReturn(Paths.get("/podcast/"));

        /* When */
        Path coverPathOf = coverBusiness.getCoverPathOf(podcast);

        /* Then */
        assertThat(coverPathOf.toString()).isEqualTo("/podcast/A Podcast/cover.jpg");
    }

    @Test
    public void should_save_cover() {
        /* Given */
        Cover cover = new Cover().setUrl("http://anUrl.jiz/foo/bar");
        UUID createdId = UUID.randomUUID();
        when(coverRepository.save(eq(cover))).thenReturn(cover.setId(createdId));

        /* When */
        Cover savedCover = coverBusiness.save(cover);

        /* Then */
        CoverAssert.assertThat(savedCover).hasId(createdId);
        assertThat(cover).isSameAs(savedCover);
        verify(coverRepository, only()).save(eq(cover));
    }
}

File: backend/src/test/kotlin/com/github/davinkevin/podcastserver/business/CoverBusinessTest.kt
package com.github.davinkevin.podcastserver.business

import com.github.davinkevin.podcastserver.service.UrlService
import com.github.tomakehurst.wiremock.WireMockServer
import com.github.tomakehurst.wiremock.client.WireMock
import com.github.tomakehurst.wiremock.core.WireMockConfiguration
import com.nhaarman.mockitokotlin2.whenever
import lan.dk.podcastserver.entity.Cover
import lan.dk.podcastserver.entity.Item
import lan.dk.podcastserver.entity.Podcast
import lan.dk.podcastserver.repository.CoverRepository
import lan.dk.podcastserver.service.properties.PodcastServerParameters
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.Spy
import org.mockito.junit.jupiter.MockitoExtension
import org.springframework.util.FileSystemUtils
import java.net.URISyntaxException
import java.nio.file.Paths
import java.util.*

/**
 * Created by kevin on 24/07/15 for Podcast Server
 */
@ExtendWith(MockitoExtension::class)
class CoverBusinessTest {

    val wireMockServer: WireMockServer = WireMockServer(WireMockConfiguration.wireMockConfig().port(PORT))

    @Mock lateinit var coverRepository: CoverRepository
    @Mock lateinit var podcastServerParameters: PodcastServerParameters
    @Spy lateinit var urlService: UrlService
    @InjectMocks lateinit var coverBusiness: CoverBusiness

    @BeforeEach
    fun beforeEach() {
        FileSystemUtils.deleteRecursively(ROOT_FOLDER.toFile())
        wireMockServer.start()
        WireMock.configureFor(PORT)
    }

    @AfterEach
    fun afterEach() = wireMockServer.stop()

    @Test
    fun `should find one`() {
        /* Given */
        whenever(coverRepository.findById(any())).then { Optional.of(Cover().apply { id = it.arguments[0] as UUID } ) }
        val id = UUID.randomUUID()
        /* When */
        val cover = coverBusiness.findOne(id)
        /* Then */
        assertThat(cover.id).isEqualTo(id)
        verify(coverRepository, times(1)).findById(id)
    }

    @Test
    fun `should say it same cover`() {
        /* Given */
        val p1 = Podcast().apply {
            cover = Cover().apply { url = "http://fakeUrl.com"  }
        }
        val p2 = Podcast().apply {
            cover = Cover().apply { url = "http://fakeurl.com"  }
        }

        /* When */
        val isSame = coverBusiness.hasSameCoverURL(p1, p2)

        /* Then */
        assertThat(isSame).isTrue()
    }

    @Test
    fun `should say its not same cover`() {
        /* Given */
        val p1 = Podcast().apply {
            cover = Cover().apply { url = "http://fakeUrl.com"  }
        }
        val p2 = Podcast().apply {
            cover = Cover().apply { url = "https://fakeurl.com"  }
        }

        /* When */
        val isSame = coverBusiness.hasSameCoverURL(p1, p2)

        /* Then */
        assertThat(isSame).isFalse()
    }

    @Test
    fun `should not download cover of this podcast due to lack of cover`() {
        /* Given */
        val podcast = Podcast()

        /* When */
        val url = coverBusiness.download(podcast)

        /* Then */
        assertThat(url).isEmpty()
    }

    @Test
    fun `should not download cover of this podcast due to empty cover url`() {
        /* Given */
        val podcast = Podcast().apply {
            cover = Cover()
        }

        /* When */
        val url = coverBusiness.download(podcast)

        /* Then */
        assertThat(url).isEmpty()
    }

    @Test
    fun `should not handle local url`() {
        /* Given */
        val podcast = Podcast().apply {
            cover = Cover().apply { url = "/url/relative" }
        }

        /* When */
        val url = coverBusiness.download(podcast)

        /* Then */
        assertThat(url).isEqualTo("/url/relative")
    }

    @Test
    fun `should download the cover of podcast`() {
        /* Given */
        val podcastTitle = "Foo"
        val imageExtension = "png"
        val defaultCoverValue = "cover"
        val podcast = Podcast().apply {
            title = podcastTitle
            id = UUID.randomUUID()
            cover = Cover().apply { url = host("/img/image.$imageExtension") }
        }

        whenever(podcastServerParameters.coverDefaultName).thenReturn(defaultCoverValue)
        whenever(podcastServerParameters.rootfolder).thenReturn(ROOT_FOLDER)

        /* When */
        val url = coverBusiness.download(podcast)

        /* Then */
        assertThat(url)
                .isEqualTo("/api/podcasts/${podcast.id}/$defaultCoverValue.$imageExtension")
    }

    @Test
    fun `should reject by network exception`() {
        /* Given */
        val podcast = Podcast().apply {
            title = "Foo"
            cover = Cover().apply { url = host("/img/image.jpg") }
        }
        whenever(podcastServerParameters.coverDefaultName).thenReturn("cover")
        whenever(podcastServerParameters.rootfolder).thenReturn(Paths.get("/tmp"))

        /* When */
        val url = coverBusiness.download(podcast)

        /* Then */
        assertThat(url).isEmpty()
    }

    @Test
    fun `should generate path cover of podcast`() {
        /* Given */
        val podcast = Podcast().apply {
                id = UUID.randomUUID()
                title = "A Podcast"
                cover = Cover().apply { url = "http://foo.bar.com/podcast/Google/aCover.jpg" }
        }

        whenever(podcastServerParameters.coverDefaultName).thenReturn("cover")
        whenever(podcastServerParameters.rootfolder).thenReturn(Paths.get("/podcast/"))

        /* When */
        val coverPathOf = coverBusiness.getCoverPathOf(podcast)

        /* Then */
        assertThat(coverPathOf.toString()).isEqualTo("/podcast/A Podcast/cover.jpg")
    }

    @Test
    fun `should save cover`() {
        /* Given */
        val cover = Cover().apply { url = "http://anUrl.jiz/foo/bar" }
        val createdId = UUID.randomUUID()
        whenever(coverRepository.save<Cover>(eq(cover))).thenReturn(cover.toBuilder().id(createdId).build())

        /* When */
        val savedCover = coverBusiness.save(cover)

        /* Then */
        assertThat(savedCover.id).isEqualTo(createdId)
        verify(coverRepository, only()).save(cover)
    }

    @Test
    fun `should not download cover of item because lack of id`() {
        /* Given */
        val item = Item()

        /* When */
        val downloaded = coverBusiness.download(item)

        /* Then */
        assertThat(downloaded).isFalse()
    }

    @Test
    fun `should not download cover of item because lack of cover`() {
        /* Given */
        val item = Item().apply {
            podcast = Podcast()
            id = UUID.randomUUID()
        }

        /* When */
        val downloaded = coverBusiness.download(item)

        /* Then */
        assertThat(downloaded).isFalse()
    }

    @Test
    fun `should download cover of item`() {
        /* Given */
        val item = Item().apply {
                id = UUID.randomUUID()
                podcast = Podcast().apply { title = "FooPodcast" }
                cover = Cover().apply { url = host("/img/image.png") }
        }
        whenever(podcastServerParameters.rootfolder).thenReturn(ROOT_FOLDER)

        /* When */
        val downloaded = coverBusiness.download(item)

        /* Then */
        assertThat(downloaded).isTrue()
    }

    @Test
    @Throws(URISyntaxException::class)
    fun `should reject by network exception for item`() {
        /* Given */
        val item = Item().apply {
            id = UUID.randomUUID()
            podcast = Podcast().apply { title = "FooPodcast" }
            cover = Cover().apply { url = host("/img/image.jpg") }
        }
        whenever(podcastServerParameters.rootfolder).thenReturn(ROOT_FOLDER)

        /* When */
        val downloaded = coverBusiness.download(item)

        /* Then */
        assertThat(downloaded).isFalse()
    }

    @Test
    fun `should find cover path of item`() {
        /* Given */
        // @formatter:off
        val item = Item().apply {
                id = UUID.randomUUID()
                cover = Cover().apply { url = "http://www.foo.bar/image.png" }
                podcast = Podcast().apply { title = "FooBarPodcast" }
        }
        whenever(podcastServerParameters.rootfolder).thenReturn(Paths.get("/tmp/"))

        /* When */
        val path = coverBusiness.getCoverPathOf(item)

        /* Then */
        assertThat(path)
                .contains(Paths.get("/tmp", "FooBarPodcast", "${item.id}.png"))
    }

    companion object {

        private val ROOT_FOLDER = Paths.get("/tmp/podcast")
        private const val PORT = 8089
        private const val HTTP_LOCALHOST = "http://localhost:$PORT"

        private fun host(path: String): String {
            return HTTP_LOCALHOST + path
        }
    }
}
